In [4]:
import pandas as pd
import numpy as np
import matplotlib as mp
import math as math
import matplotlib.pyplot as plt



In [5]:
# set the kidney function
def kidney(x, y):
    return (x**2 + y**2)**2 - (x**3 + y**3)

# set the disc function
def disc(x, y):
    return (x - 0.25)**2 + (y - 0.25)**2 - 0.125

# define rectangle method
def rectangle_method(x_min, x_max, y_min, y_max, n):
    # set the intervals to check for the x and y distances (since kidney is not a proper function, we need to check all directions for interesections in small intervals) 
    dx = (x_max - x_min) / n
    dy = (y_max - y_min) / n
    area = 0.0
    
    # for every x, check all y, and check to see if we interesect the kidney function, without being in the disk, if so add the area 
    for i in range(n):
        for j in range(n):
            # use rectangular method and get middle of the interval
            x = x_min + (i + 0.5) * dx
            y = y_min + (j + 0.5) * dy
            # if the x and y do not lie in the disk, and is within kidney, add the area to area counter
            if kidney(x, y) <= 0 and disc(x, y) > 0:
                area += dx * dy
    
    return area

def trapezoid_method(x_min, x_max, y_min, y_max, n):
    # set the intervals to check for the x and y distances (since kidney is not a proper function, we need to check all directions for interesections in small intervals) 
    dx = (x_max - x_min) / n
    dy = (y_max - y_min) / n
    area = 0.0

    # for every x, check all y, and check to see if we interesect the kidney function, without being in the disk, if so add the area 
    for i in range(n):
        for j in range(n):
            # using trapezoidal method, get both sides of the interval
            x1 = x_min + i * dx
            y1 = y_min + j * dy
            x2 = x1 + dx
            y2 = y1 + dy
            
            #make sure all points in the trapezoid lie within the area between the disk and the kidney
            f1 = kidney(x1, y1) <= 0 and disc(x1, y1) > 0
            f2 = kidney(x2, y1) <= 0 and disc(x2, y1) > 0
            f3 = kidney(x1, y2) <= 0 and disc(x1, y2) > 0
            f4 = kidney(x2, y2) <= 0 and disc(x2, y2) > 0
            
            # calculate area 
            area += (f1 + f2 + f3 + f4) * dx * dy / 4
    
    return area


In [6]:
x_min, x_max = -1, 1
y_min, y_max = -1, 1
# # of intervals to be used
n = 1000

rectangle_area = rectangle_method(x_min, x_max, y_min, y_max, n)
trapezoidal_area = trapezoid_method(x_min, x_max, y_min, y_max, n)

print("Area using rectangle:", rectangle_area)
print("Area using trapezoid:", trapezoidal_area)


Area using rectangle: 0.5890480000005064
Area using trapezoid: 0.5891240000004702


The area using rectangle is 0.5890

The area using trapezoid is 0.5891

In [7]:
def problem_2(N):
    # generate a NxN matrix using random uniform dist with a = -1 and b = 1 
    A = np.random.uniform(-1, 1, (N, N))

    # create a vector of 1s
    b = np.ones((N, 1))

    # compute the inverse of A
    inv = np.linalg.inv(A)

    # solve for x
    x = np.dot(inv, b)

    return x


N = 66

x= problem_2(N)

print("solution is: \n", x)

solution is: 
 [[ 0.89685001]
 [ 0.03964257]
 [-0.12080923]
 [-1.64687377]
 [-0.32850757]
 [-2.00858128]
 [ 0.06311924]
 [ 0.26819789]
 [ 0.26701171]
 [-0.28580902]
 [-0.64177615]
 [-0.88039224]
 [-0.94378623]
 [ 1.454253  ]
 [ 0.20225785]
 [-0.14428742]
 [ 0.42469619]
 [-1.04986317]
 [ 1.19483823]
 [ 0.89776282]
 [-0.88760474]
 [-0.96040869]
 [ 0.57740279]
 [ 1.11256017]
 [-0.10250896]
 [-0.6489052 ]
 [ 0.04485481]
 [ 0.92800279]
 [ 0.88914871]
 [ 0.18553437]
 [-0.12889894]
 [ 0.67228974]
 [ 0.08361874]
 [ 0.23224393]
 [-0.17456244]
 [-0.60614631]
 [-1.24237205]
 [ 0.28571104]
 [ 0.82626025]
 [-0.77037613]
 [ 0.06824107]
 [-0.36212748]
 [ 1.02867612]
 [-1.06075961]
 [ 0.76850574]
 [ 1.61627858]
 [-0.32501129]
 [ 1.54070227]
 [ 1.75292294]
 [ 1.79221729]
 [-0.28136305]
 [-0.45203962]
 [ 0.8611815 ]
 [ 0.15328934]
 [-0.84396787]
 [-0.95772434]
 [ 0.74593611]
 [ 1.77625115]
 [-0.61743374]
 [ 0.3679868 ]
 [-0.44489966]
 [-1.84237572]
 [ 0.05964499]
 [ 0.12690022]
 [-0.78524115]
 [-0.01280

solution is: 
 [[ 0.89685001]
 [ 0.03964257]
 [-0.12080923]
 [-1.64687377]
 [-0.32850757]
 [-2.00858128]
 [ 0.06311924]
 [ 0.26819789]
 [ 0.26701171]
 [-0.28580902]
 [-0.64177615]
 [-0.88039224]
 [-0.94378623]
 [ 1.454253  ]
 [ 0.20225785]
 [-0.14428742]
 [ 0.42469619]
 [-1.04986317]
 [ 1.19483823]
 [ 0.89776282]
 [-0.88760474]
 [-0.96040869]
 [ 0.57740279]
 [ 1.11256017]
 [-0.10250896]
 [-0.6489052 ]
 [ 0.04485481]
 [ 0.92800279]
 [ 0.88914871]
 [ 0.18553437]
 [-0.12889894]
 [ 0.67228974]
 [ 0.08361874]
 [ 0.23224393]
 [-0.17456244]
 [-0.60614631]
 [-1.24237205]
 [ 0.28571104]
 [ 0.82626025]
 [-0.77037613]
 [ 0.06824107]
 [-0.36212748]
 [ 1.02867612]
 [-1.06075961]
 [ 0.76850574]
 [ 1.61627858]
 [-0.32501129]
 [ 1.54070227]
 [ 1.75292294]
 [ 1.79221729]
 [-0.28136305]
 [-0.45203962]
 [ 0.8611815 ]
 [ 0.15328934]
 [-0.84396787]
 [-0.95772434]
 [ 0.74593611]
 [ 1.77625115]
 [-0.61743374]
 [ 0.3679868 ]
 [-0.44489966]
 [-1.84237572]
 [ 0.05964499]
 [ 0.12690022]
 [-0.78524115]
 [-0.01280392]]